File: typescript/src/model/sections/wdl-metadata-base.ts
/** Metadata section nodes. */
import { WdlStringKeyValue } from '../base/wdl-key-value.js';
import type { WdlNode } from '../base/wdl-node.js';
import type { WdlExpression } from '../expressions/wdl-expression.js';

/** Single metadata entry such as `description: "..."`. */
export class WdlMetadataEntry extends WdlStringKeyValue {
  /** Creates a metadata entry from its key and optional expression value. */
  public constructor(key?: string, value?: WdlExpression) {
    super(key, value);
  }
}

/** Base container for ordered metadata entries. */
export class WdlMetadataBase implements WdlNode {
  private readonly elementValues: WdlMetadataEntry[] = [];
  /** Returns the ordered metadata entries in the section. */
  public elements(): WdlMetadataEntry[] {
    return this.elementValues;
  }
}

/** General metadata section. */
export class WdlMetadata extends WdlMetadataBase {}

/** Parameter metadata section. */
export class WdlParameterMetadata extends WdlMetadataBase {}

File: python/src/wdl_model/model/sections/wdl_metadata_base.py
"""Metadata section nodes for tasks, workflows, and structs."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from wdl_model.model.base import WdlNode, WdlStringKeyValue
from wdl_model.model.expressions import WdlExpression


class WdlMetadataEntry(WdlStringKeyValue):
    """Single metadata entry such as ``description: \"text\"``."""

    def __init__(self, key: str | None = None, value: WdlExpression | None = None):
        super().__init__(key, value)


@dataclass
class WdlMetadataBase(WdlNode):
    """Base container for ordered metadata entries."""

    _elements: deque[WdlMetadataEntry] = field(default_factory=deque)

    def elements(self) -> deque[WdlMetadataEntry]:
        """Return the ordered metadata entries contained in the section."""
        return self._elements


class WdlMetadata(WdlMetadataBase):
    """General metadata section."""

    pass


class WdlParameterMetadata(WdlMetadataBase):
    """Parameter metadata section."""

    pass
